gate-web/src/main/groovy/com/netflix/spinnaker/gate/controllers/ApiExtensionController.java

```diff
@@ -0,0 +1,81 @@
+package com.netflix.spinnaker.gate.controllers;
+
+import static org.springframework.web.bind.annotation.RequestMethod.*;
+
+import com.netflix.spinnaker.gate.api.extension.ApiExtension;
+import com.netflix.spinnaker.gate.api.extension.HttpRequest;
+import com.netflix.spinnaker.gate.api.extension.HttpResponse;
+import com.netflix.spinnaker.kork.annotations.Alpha;
+import com.netflix.spinnaker.kork.exceptions.SpinnakerException;
+import com.netflix.spinnaker.kork.web.exceptions.NotFoundException;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.servlet.http.HttpServletRequest;
+import okhttp3.internal.http.HttpMethod;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * A top-level {@code RestController} that exposes all api extensions under a common
+ * `/extensions/{{ApiExtension.id}}` path.
+ */
+@Alpha
+@RestController
+@RequestMapping("/extensions")
+public class ApiExtensionController {
+  private List<ApiExtension> apiExtensions;
+
+  @Autowired
+  public ApiExtensionController(List<ApiExtension> apiExtensions) {
+    this.apiExtensions = apiExtensions;
+  }
+
+  @RequestMapping(value = {"/{extension}/**"})
+  public ResponseEntity<Object> any(
+      @PathVariable(value = "extension") String extension,
+      @RequestParam Map<String, String> requestParams,
+      HttpServletRequest httpServletRequest) {
+    HttpRequest httpRequest =
+        HttpRequest.of(
+            httpServletRequest.getMethod(),
+            httpServletRequest.getRequestURI().replace("/extensions/" + extension, ""),
+            requestParams);
+
+    if (HttpMethod.permitsRequestBody(httpRequest.getMethod())) {
+      try {
+        httpRequest.setBody(
+            httpServletRequest
+                .getReader()
+                .lines()
+                .collect(Collectors.joining(System.lineSeparator())));
+      } catch (IOException e) {
+        throw new SpinnakerException("Unable to read request body", e);
+      }
+    }
+
+    Optional<ApiExtension> apiExtension =
+        apiExtensions.stream()
+            .filter(e -> e.id().equalsIgnoreCase(extension) && e.handles(httpRequest))
+            .findFirst();
```

Comment:
I think we should log a message if more than one `ApiExtension` handles a particular `httpRequest`, also logging which one it is going to choose.

nit: `findFirst` is a little too non-deterministic for my tastes in the case of a collision. I'm torn between two solutions: `ApiExtension` supporting priority ordering, or being strict and throwing an exception if any collision occurs.

Reply:
I'd like to err on the side of just throwing the exception -- I recall we're doing that for stage type resolution?

Reply:
Yes, we just explode if we can't find a stage.